framework/src/org/checkerframework/framework/type/AsSuperVisitor.java

```diff
@@ -0,0 +1,813 @@
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedDeclaredType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedIntersectionType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedNullType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedPrimitiveType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedTypeVariable;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedUnionType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedWildcardType;
+import org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor;
+import org.checkerframework.framework.util.AnnotatedTypes;
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.InternalUtils;
+import org.checkerframework.javacutil.TypesUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Types;
+
+/**
+ * Implements asSuper {@link AnnotatedTypes#asSuper(AnnotatedTypeFactory, AnnotatedTypeMirror,
+ * AnnotatedTypeMirror)}
+ */
+public class AsSuperVisitor extends AbstractAtmComboVisitor<AnnotatedTypeMirror, Void> {
+
+    private final Types types;
+    private final AnnotatedTypeFactory annotatedTypeFactory;
+
+    public AsSuperVisitor(AnnotatedTypeFactory annotatedTypeFactory) {
+        this.annotatedTypeFactory = annotatedTypeFactory;
+        types = annotatedTypeFactory.types;
+    }
+
+    /**
+     * Implements asSuper.
+     * See {@link AnnotatedTypes#asSuper(AnnotatedTypeFactory, AnnotatedTypeMirror,
+     * AnnotatedTypeMirror)} for details.
+     *
+     * @param type      Type from which to copy annotations
+     * @param superType a type whose erased Java type is a supertype of {@code type}'s erased
+     *                  Java type.
+     * @return A copy of {@code superType} with annotations copied from {@code type} and type
+     * variables substituted from {@code type}.
+     */
+    @SuppressWarnings("unchecked")
+    public <T extends AnnotatedTypeMirror> T asSuper(AnnotatedTypeMirror type, T superType) {
+        if (type == null || superType == null) {
+            ErrorReporter.errorAbort("AsSuperVisitor type and supertype cannot be null.");
+            return null; // dead code
+        } else if (type == superType) {
+            return (T) type.deepCopy();
+        }
+
+        // This visitor modifies superType and may return type, so pass it copies so that it that
+        // parameters to asSuper are not changed and a copy is returned.
+        AnnotatedTypeMirror copyType = type.deepCopy();
+        AnnotatedTypeMirror copySuperType = superType.deepCopy(false);
```

Comment:
I deleted the following comment for @wmdietl on this line:
>Can you comment why the visitor gets copies of the parameters? Does the visitor modify the parameters?
Why is one a shallow copy and the other not? Can you add comments?
Extract the type.deepCopy into another local variable.